### ISSUE TITLE
carthage update dependency version higher than i specify

### ISSUE THREAD

**Issue by zhongWJ** [Carthage/Carthage#2516]
* carthage install method: [ ] .pkg, [√ ] homebrew, [ ] source
* `which carthage`: /usr/local/bin/carthage
* `carthage version`: 0.30.1
* `xcodebuild -version`: Xcode 10.0 Build version 10L201y
* Are you using `--no-build`?  No
* Are you using `--no-use-binaries`?  No
* Are you using `--use-submodules`?  No
* Are you using `--cache-builds`?  No
* Are you using `--new-resolver`?  No

**Cartfile**
```
github "apache/incubator-weex" == 0.18.0
```

**Carthage Output**
```
*** Fetching incubator-weex
*** Checking out incubator-weex at "0.18.0.1"
```
I specify the version to exactly 0.18.0, but when i use `carthage update`, it checkout a higher version 0.18.0.1. It seems to use compatible version.

**Comment by zhongWJ** [comment]
I change my Cartfile to this:

```
github "apache/incubator-weex" "0.18.0"
```
This time carthage works as expected.

**Comment by blender** [comment]
Have you tried with --new-resolver?

**Comment by zhongWJ** [comment]
carthage update --new-resolver
*** Fetching incubator-weex
*** Checking out incubator-weex at "0.18.0.1"
i tried with --new-resolver, don't work.

**Comment by marcoconti83** [comment]
`0.18.0.1` is not a valid semantic version. 

More details here: https://semver.org/

**Comment by marcoconti83** [comment]
@mdiep I think the code still implements the behavior that if a semver version is requested, but there is a non-semver version, it will pick the non-semver version (a random one!).

This is one point that I tried to address in my original PR more than one year ago but what I understood from your [comment](https://github.com/Carthage/Carthage/pull/1553#issuecomment-258733536) is that this would break compatibility and you would prefer to leave the current behavior.

If this is not the case, I would be happy to change the code so that it will ignore any non-semver version and (in this example by @zhongWJ ) return an error because it finds no version that can satisfy the requirement.

**Comment by mdiep** [comment]
@marcoconti83 It's more subtle than that. We should only pick a non-semver version if explicitly requested. (i.e. it appears in a `Cartfile`) The issue here is that we're picking one when it hasn't been requested.

**Comment by jdhealy** [comment]
@marcoconti83 @mdiep — to note (and maybe clarify), `SemanticVersion.from(Scanner(string: "4.3.2.1"))` does return a `Result.success`¹.

So, while not being technically [SemVer](https://semver.org/), it will get parsed into a `CarthageKit.SemanticVersion` for [versions dating back to 2015 (at least)](https://github.com/Carthage/Carthage/pull/508#issuecomment-111744540).

We [scan characters from `versionCharacterSet`](https://github.com/Carthage/Carthage/blob/0.29.0/Source/CarthageKit/Version.swift#L40-L41), [split on full stops (omitting empty subsequences)](https://github.com/Carthage/Carthage/blob/0.29.0/Source/CarthageKit/Version.swift#L73), [require at least 2 integers](https://github.com/Carthage/Carthage/blob/0.29.0/Source/CarthageKit/Version.swift#L78-L90), but perform no check that a full stop doesn’t immediately follow the patch version.

While this code did get reevaluated to some degree during the review process for #2412 — code like `SemanticVersion.from(Scanner(string: "4....3.2"))` was considered — it’s somewhat of an oversight that we didn’t implement a stricter checking on whether post-patch-version characters exists in [what we split out of `versionCharacterSet`](https://github.com/Carthage/Carthage/blob/0.30.1/Source/CarthageKit/Version.swift#L119-L120).

For post-patch-version characters that don’t fall into [`versionCharacterSet`](https://github.com/Carthage/Carthage/blob/0.29.0/Source/CarthageKit/Version.swift#L40-L41), [`scanner.isAtEnd`](https://github.com/Carthage/Carthage/blob/0.30.1/Source/CarthageKit/Version.swift#L142-L144) performs that check.

⑴ `expect(SemanticVersion.from(Scanner(string: "4.3.2.1")).value) == SemanticVersion(major: 4, minor: 3, patch: 2)` will hold true for Carthage versions 0.30.1, 0.29.0, and many previous versions.

**Comment by mdiep** [comment]
Ah, okay. Then this can be solved without considering non-semantic versions.

We definitely should tighten that constraint to reject `0.18.0.1` as a semantic version.

**Comment by marcoconti83** [comment]
@jdhealy thanks for debugging this. It’s something I missed during my PR then, I’ll fix it and open a PR.

**Comment by marcoconti83** [comment]
Tentative fix: https://github.com/Carthage/Carthage/pull/2518

**Comment by blender** [comment]
@zhongWJ  please try `brew install carthage --HEAD`

**Comment by zhongWJ** [comment]
@blender it works!Thx!

**Comment by ikesyo** [comment]
Fixed by #2518.
